### ISSUE TITLE
virtual attributes are always return by model

### ISSUE THREAD

**Issue by p2227** [sequelize/sequelize#5566]
```
var pinyin = require('pinyin');

module.exports = function(sequelize, DataTypes) {
  var moneyType = sequelize.options.dialect === 'mssql' ? 'MONEY' : DataTypes.INTEGER;
  return sequelize.define('Pt', {
    usertype:{
      type:DataTypes.VIRTUAL,
      get(){
        return '0'
      }
    },
    pinyin:{
      type:DataTypes.VIRTUAL,
      scopes:['user'],
      get(){
        return pinyin(this.getDataValue('TX_Name'),{
          segment:false,
          style:pinyin.STYLE_NORMAL
        }).join('')
      }
    },
    ID_User: {
      primaryKey: true,
      type: DataTypes.STRING,
      allowNull: false
    },
    TX_Name: {
      type: DataTypes.STRING,
      allowNull: true
    },
    DT_Birth: {
      type: DataTypes.DATE,
      get:function(){
        var d = this.getDataValue('DT_Birth')
        return d ? d.toLocaleString() : d; 
      },
      allowNull: true
    },
    TX_IC: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_IMG: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_Location: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_Signature: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_Username: {
      type: DataTypes.STRING,
      allowNull: false
    },
    TX_Password: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_Contact: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_QQ: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_WeChat: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_QR: {
      type: DataTypes.STRING,
      allowNull: true
    },
    DT_Delivery: {
      type: DataTypes.DATE,
      get:function(){
        var d = this.getDataValue('DT_Delivery')
        return d ? d.toLocaleString() : d; 
      },
      allowNull: true
    },
    TX_Name2: {
      type: DataTypes.STRING,
      allowNull: true
    },
    TX_Contact2: {
      type: DataTypes.STRING,
      allowNull: true
    },
    MN_Credit: {
      type: moneyType,
      allowNull: true
    },
    MN_Deposit: {
      type: moneyType,
      allowNull: true
    },
  }, {
    defaultScope: {
      attributes: { exclude: ['TX_Password'] }
    },
    scopes:{
      'user':{
        attributes:['ID_User','TX_Name','DT_Birth','TX_IC','TX_IMG','TX_Location','TX_Signature','TX_Username','TX_Password','TX_QR','TX_Contact','TX_QQ','TX_WeChat']
      }
    },
    updatedAt:'DTTM_Updated',
    createdAt:'DTTM_Created',
    tableName: 'T_Pt',
    freezeTableName: true
  });
};
```

the attributes usertype and pinyin are always return where i findOne/findAll event i use model.scope('xxx')

**Comment by wkkwok** [comment]
+1
Virtaul attributes seems to be added last, can't remove them even with afterFind Hook.

**Comment by wkkwok** [comment]
```
this.addHook('afterFind', 'removeActivationCode', async function(users, options) {
	if (!users)
		return;
	if (users.constructor != Array)
		users = [users];
	for (var i = 0; i < users.length; i++) {
		if (users[i]._hasCustomGetters) {
			users[i]._customGetters = _.pick(users[i]._customGetters, options.attributes)
		}
	}
});
```
I finally figure it out myself. It seems that custom getter are called only in` Instance.get` after the all options are parsed. Therefore,` find(options.attributes[])` will not affect the output. I am not sure if this is the intended outcome or a bug.

**Comment by zuck** [comment]
In my opinion this is a bug that must be fixed.

**Comment by janmeier** [comment]
@zuck Contributions welcome :) https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md

**Comment by nstuyvesant** [comment]
Would it make sense to update the docs for now to qualify
```
Model.findAll({
  attributes: { exclude: ['baz'] }
});
```
works but only if baz is not type: DataTypes.VIRTUAL?

**Comment by javaguy** [comment]
I looked into this issue as well since I'm facing the same problem.  I believe this can be fixed in Instance.get which indiscriminately calls all the custom getters. Instead there should probably be a check to see if the getter corresponds to an included attribute. 

@janmeier I'd be happy to submit a PR for this if you can confirm I'm on the right track.

**Comment by huynhj93** [comment]
@wkkwok Is that fix now deprecated? I tried implementing it, but options.attributes seems to exclude the virtual attributes now. Thus, your model._customGetters will be an empty object now.

**Comment by scottbeacher** [comment]
This is still an issue.

**Comment by vinhboy** [comment]
@javaguy I was able to fix it in my environment by patching that file with

https://github.com/sequelize/sequelize/blob/v3.30.4/lib/instance.js#L207

```javascript
    if (this._hasCustomGetters) {
      for (_key in this._customGetters) {
        if (this.$options.attributes && !this.$options.attributes.includes(_key)) {
          continue;
        }
        if (this._customGetters.hasOwnProperty(_key)) {
          values[_key] = this.get(_key, options);
        }
      }
    }
```
You should submit a PR if you have time to work on. Thanks.

**Comment by Undre4m** [comment]
+1

**Comment by onerciller** [comment]
+1

**Comment by CherryChann** [comment]
This is still issue and I cannot exclude the attributes with getter function..:(

**Comment by chemic** [comment]
Is it still like that, or there's a way to filter VIRTUAL out somehow?
Thanks!

**Comment by smithaitufe** [comment]
So basically, how are virtual field ignored?

**Comment by elwin1234** [comment]
This issue is still not solved in current stable release `v4.42.0`, right? **Any** virtual property, whether it has a custom get() function or not, is never correctly excluded when using `findByPk`, `findAll`, `find*` functions?

**Comment by arthurfauq** [comment]
Any update on this issue ? Still not solved in release `v5.21.3`...
